device/common/include/traccc/finding/device/impl/build_tracks.ipp

```diff
@@ -86,64 +60,45 @@ TRACCC_DEVICE inline void build_tracks(const global_index_t globalIndex,
             L = links[L.previous.first][L.previous.second];
         }
 
-        // Break if the measurement is still invalid
-        if (L.meas_idx >= measurements.size()) {
-            success = false;
-            break;
-        }
+        assert(L.meas_idx < measurements.size());
 
-        *it = {measurements.at(L.meas_idx)};
-        num_inserted++;
+        final_candidates.at(globalIndex).items.at(i) =
+            measurements.at(L.meas_idx);
 
         // Sanity check on chi2
         assert(L.chi2 < std::numeric_limits<traccc::scalar>::max());
         assert(L.chi2 >= 0.f);
 
-        ndf_sum += static_cast<scalar>(it->meas_dim);
+        ndf_sum += static_cast<scalar>(measurements.at(L.meas_idx).meas_dim);
         chi2_sum += L.chi2;
 
         // Break the loop if the iterator is at the first candidate and fill the
         // seed and track quality
-        if (it == cands_per_track.rend() - 1) {
-            seed = seeds.at(L.previous.second);
-            trk_quality.ndf = ndf_sum - 5.f;
-            trk_quality.chi2 = chi2_sum;
-            trk_quality.n_holes = L.n_skipped;
+        if (i == 0) {
+            seed_idx = L.previous.second;
+            final_n_skipped = L.n_skipped;
         } else {
             L = links[L.previous.first][L.previous.second];
         }
     }
 
+    final_candidates.at(globalIndex).header.seed_params = seeds.at(seed_idx);
+    final_candidates.at(globalIndex).header.trk_quality.chi2 = chi2_sum;
+    final_candidates.at(globalIndex).header.trk_quality.ndf = ndf_sum - 5.f;
+    final_candidates.at(globalIndex).header.trk_quality.n_holes =
+        final_n_skipped;
```

Comment:
I honestly don't know this. Since the compiler might already do this as efficiently as possible.

But would it not be more efficient to write this as:

```c++
final_candidates.at(globalIndex).header = {...};
```

? Note that I'm very aware that I'm also talking against my SoA efforts with this. 😛 Since if this formalism is indeed more efficient, the SoA migration will make this part worse. 😦

Reply:
Mwah, in this case the efficiency difference is going to be so small that it is not worth caring about. That said I do prefer the syntax with aggregate notation, so I'll use that. :+1: